File: 백준/Silver/1926. 그림/그림.py
import sys
from collections import deque

input = sys.stdin.readline


def bfs(i, j):
    cnt = 1
    q = deque()
    q.append([i, j])
    visited[i][j] = True
    while q:
        x, y = q.popleft()
        
        for k in range(4):
            nx = x + dx[k]
            ny = y + dy[k]
            
            if 0 <= nx < n and 0 <= ny < m and graph[nx][ny] == 1 and visited[nx][ny] is False:
                cnt += 1
                visited[nx][ny] = True
                q.append([nx, ny])
    return cnt


n, m = map(int, input().split())
graph = []
for _ in range(n):
    graph.append(list(map(int, input().split())))
visited = [[False] * m for _ in range(n)]

dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]

pictures = 0
max_cnt = 0

for i in range(n):
    for j in range(m):
        if graph[i][j] == 1 and visited[i][j] is False:
            pictures += 1
            max_cnt = max(max_cnt, bfs(i, j))


print(pictures)
print(max_cnt)

File: 백준/Silver/1926. 그림/그림.js
const path = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(path).toString().trim().split("\n");

const directions = [
  [0, 1],
  [1, 0],
  [0, -1],
  [-1, 0],
];

function solution(input) {
  const [n, m] = input[0].split(" ").map(Number);

  let count = 0; // 그림의 개수
  let maxArea = 0; // 가장 넓은 그림의 넓이

  const graph = [];
  for (let i = 1; i < n + 1; i++) {
    const row = input[i].split(" ").map(Number);
    graph.push(row);
  }

  const visited = Array.from({ length: n }, () => Array(m).fill(false));

  function dfs(x, y) {
    let area = 1; // 현재 그림의 넓이
    const stack = [[x, y]]; // 스택을 사용한 DFS

    visited[x][y] = true;

    while (stack.length) {
      const [cx, cy] = stack.pop();

      for (const [dx, dy] of directions) {
        const nx = cx + dx;
        const ny = cy + dy;

        if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
        if (graph[nx][ny] === 0 || visited[nx][ny]) continue;

        visited[nx][ny] = true;
        stack.push([nx, ny]);
        area++;
      }
    }

    return area;
  }

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (graph[i][j] === 1 && !visited[i][j]) {
        const area = dfs(i, j);
        maxArea = Math.max(maxArea, area);
        count++;
      }
    }
  }

  console.log(count);
  console.log(maxArea);
}

solution(input);
